### ISSUE TITLE
Misleading error message `Uncaught Invariant Violation: RelayQueryNode: Abstract class cannot be instantiated.`

### ISSUE THREAD

**Issue by rosskevin** [facebook/relay#1368]
`Uncaught Invariant Violation: RelayQueryNode: Abstract class cannot be instantiated.`

I first tried discord and was directed here.

**TLDR;**  - how do I go about diagnosing the cause of this error?

---
## Now, the long story

BTW - this is not a decorator issue (which is what comes up on google).  The code in question works as a single project, breaks when split into two projects.

I have a webpack/babel bundling/transpiling process that was working well. As we grew, I wanted to split from a single project to multiple for reuse e.g. `app` vs `lib`.  Therein lies the problem.  I tried to include the `lib` source via webpack (which I've done before and seems ok), but suddenly I'm experiencing the error above.  If I first transpile the lib to ES5, then use it, I don't get the error.  I don't want to build to ES5 first for a number of reasons, primarily because the `lib` in question is the company's core (white label), and flattening it out as a single module (for exports) has name conflicts.

Here are some possibilities I have considered but not necessarily ruled out:
### Possiblity 1

I don't know how path may affect the `babel-relay-plugin`, but the path change provides the potential this is the the problem i.e. resolution of types.  Almost all code still resides in the `lib` (including relay all queries/containers etc), and the `app` at this point is as thin as possible.  To counter this, I added code to `babel-relay-plugin-loader` and compared the generated output of working vs non-working, and there is no difference.  Perhaps this isn't enough to exclude this possibility, or perhaps it is.
### Possibility 2

We have shrunk the scope of `app`'s use of `lib` down to a minimum.  We first rendered the component without a relay container successfully. 

Then we added back a single relay container/query.  We see:
- no query on the server side
- no content rendered from the container component
- error in the log: `Uncaught Invariant Violation: RelayQueryNode: Abstract class cannot be instantiated.`

**I'm stuck without some help.**

(EDIT I fixed the ref issue since my initial posting of this issue)
/cc @bmcdaniel11

**Comment by rosskevin** [comment]
We have determined that like `react`, `react-relay` does not behave when it is found in more than just the root `node_modules`.  For example, the `lib` relies on `react-relay` as a `dependency`, and so does the `app`.

I don't know the reason react or relay behave this way, but I do know that the [react library is proactive to raise an error to a note stating what the causes may be](https://facebook.github.io/react/warnings/refs-must-have-owner.html).  So, two things:

1. If someone could point me to anything that would enlighten me as to why react/relay do this, I'm interested in learning.  Any thoughts/pointers are welcome.

2. I think not having a relevant error message is a legitimate bug in relay.  We spent 1.5 days trying to figure it out, and I would very much like it if there was, at a minimum, an error message similar to that of react.

**Comment by rosskevin** [comment]
I've found some background info on what to do for reusable libraries and react (which advice I assume also applies for relay): http://stackoverflow.com/a/30454133/2363935

**Comment by josephsavona** [comment]
Thanks for debugging this and providing more info. Having multiple copies of `react-relay` within an application would definitely explain this bug; the duplication of modules this is a deficiency of npm that libraries just have to work around.

We'd be open to a PR to add a more extensive error message here!

**Comment by wincent** [comment]
Going through and closing issues that haven't had any recent activity. We'd still be open to reviewing a PR that improved the error messaging here, but it's not something that any of us are actively working on internally, so I am going to close this. Thanks very much for reporting it in such detail though!
